src/erlastic_search.erl

```diff
@@ -664,21 +697,33 @@ commas([]) ->
 commas([H | T]) ->
     << H/binary, << <<",", B/binary>> || B <- T >>/binary >>.
 
--spec bulk_index_docs_header(binary(), binary(), binary(), list()) -> binary().
-bulk_index_docs_header(Index, Type, Id, HeaderInformation) ->
-    IndexHeaderJson1 = [
-        {<<"_index">>, Index}
-        ,{<<"_type">>, Type}
-        | HeaderInformation
+build_header(Operation, Index, Type, Id, HeaderInformation) ->
+    Header1 = [
+      {<<"_index">>, Index}
+      | HeaderInformation
     ],
 
-    IndexHeaderJson2 = case Id =:= undefined of
-        true ->  IndexHeaderJson1;
-        false -> [ {<<"_id">>, Id} | IndexHeaderJson1]
-    end,
-
-    %% we cannot use erls_json to generate this, see the doc string for `erls_json:encode/1'
-    jsx:encode([{<<"index">>, IndexHeaderJson2}]).
+    Header2 = case Type =:= undefined of
+                true -> Header1;
+                false -> [{<<"_type">>, Type} | Header1]
+              end,
+
+    Header3 = case Id =:= undefined of
+                true -> Header2;
+                false -> [{<<"_id">>, Id} | Header2]
+              end,
+
+    jsx:encode([{erlang:atom_to_binary(Operation, utf8), Header3}]).
+
+build_body(delete, no_body) ->
+    [<<"\n">>];
+build_body(update, Doc) ->
+    DocBin = maybe_encode_doc(Doc),
+    Json = <<"{\"doc\":", DocBin/binary, "}">>,
+    [<<"\n">>, Json, <<"\n">>];
+build_body(_Operation, Doc) ->
+    Json = maybe_encode_doc(Doc),
+    [<<"\n">>, Json, <<"\n">>].
```

Comment:
Nit-picking at this point, but this fun could be written as:
```
build_body(delete, no_body) ->
  [];
build_body(Operation, Doc) ->
  [maybe_wrap_for_update(Operation, maybe_encode_doc(Doc)), <<"\n">>].

maybe_wrap_for_update(update, EncodedDoc) -> <<"{\"doc\":", DocBin/binary, "}">>;
maybe_wrap_for_update(_Else, EncodedDoc) -> EncodedDoc.
```
then line 679 above becomes `Header ++ [<<"\n">> | build_body(Operation, Doc)]`

But meh, not necessarily much better now that I've written it. Your call :)